### ISSUE TITLE
Consider flattening allOf + 1 subschema as a way to add descriptions to properties

### ISSUE THREAD

**Issue by czechboy0** [apple/swift-openapi-generator#31]
OpenAPI 3.0.x doesn't support adding descriptions to object properties, only to schemas.

That's not a problem when the schema is an inline one, as the description of the inline schema simply becomes the description of the property (works today).

However, when the property references a schema in `#/components/schemas/...`, there's no room to put a description on the property.

For example, consider:

```yaml
components:
  schemas:
    Route:
      type: object
      properties:
        from:
          $ref: '#components/schemas/Location'
        to:
          $ref: '#components/schemas/Location'
      required:
        - from
        - to
```

Above, one might reasonably want to document the `from` and `to` properties differently, but can't in OpenAPI 3.0 (however, it is possible to add descriptions to `$ref` schemas in OpenAPI 3.1).

A commonly used workaround, even [recommended](https://github.com/OAI/OpenAPI-Specification/issues/2033#issuecomment-543973334) by the JSON schema contributors, is to wrap the reference in an `allOf` inline schema, and add the property there, making it:

```yaml
components:
  schemas:
    Route:
      type: object
      properties:
        from:
          description: The location from which the traveler departs.
          allOf:
            - $ref: '#components/schemas/Location'
        to:
          description: The location to which the traveler arrives.
          allOf:
            - $ref: '#components/schemas/Location'
      required:
        - from
        - to
```

Today, the generator produces a struct for the `allOf`, meaning that all access to the nested schema is with an extra property access.

Since this is an officially recommended workaround for a design issue in OpenAPI, we could have the generator flatten that allOf level when it only contains a single child schema, and only use the description from the all of, resulting in being able to add a comment on a property that references a reusable type.

The main question here is:
1. Should it be default behavior?
2. Should it be possible to opt-out of this shortcut?
3. Should it be opt-in only?

**Comment by liamnichols** [comment]
I agree with your proposal in the comment above @czechboy0 👍 

From my time working on CreateAPI, I don’t think that I ever saw a reason to want to opt out of this behaviour.

This is particularly useful for making properties to reference schemas nullable too 🙏

**Comment by czechboy0** [comment]
Thinking about this a bit more, maybe we don't need this, here's an alternative idea:

as we embrace the design principles of this project outlined in [Project scope and goals](https://swiftpackageindex.com/apple/swift-openapi-generator/0.1.3/documentation/swift-openapi-generator/project-scope-and-goals), we need to avoid bringing in functionality that doesn't _have to be_ in the generator, and could be achieved by a tool that operates either on the source OpenAPI document, or on the generated Swift code.

In earlier comments above, I considered adding descriptions to object properties that are references as _not covered_ by the default _out of scope_ policy, because I was only thinking about OpenAPI 3.0.x, where it really is impossible, and the official recommendation as a workaround is to wrap the property in an `allOf`. Unfortunately, that leads to an extra property access, going from 1 to 2, which is less than ideal.

_However_, OpenAPI 3.1 does support adding descriptions to reference schemas, so the natural recommendation should be: use OpenAPI 3.1.

Originally, I wasn't sure how quickly we would be able to get to supporting 3.1, but based on recent discussions in #39, and OpenAPIKit now having alpha8, which allows us to migrate piece-meal, I think we should just prioritize supporting OpenAPI 3.1 as a way to resolve this issue without additional changes.

**Comment by czechboy0** [comment]
Blocking on OpenAPI 3.1 support for now, after that's done, we'll be in a better position to decide what to do with this.

**Comment by czechboy0** [comment]
Ok, when this PR lands, this will get closed - in OpenAPI 3.1, one can attach `description` (and other core properties, like `deprecated`) directly to a `$ref` schema as well, removing the need for this workaround.

Test here: https://github.com/apple/swift-openapi-generator/pull/219/commits/e669276c2ca1db217da24de38b2f0f098459cb1b

**Comment by czechboy0** [comment]
Now that we support OpenAPI 3.1, descriptions on refs are supported. So the solution here is to migrate your document to 3.1, then descriptions will work.
